File: 백준/Silver/11053. 가장 긴 증가하는 부분 수열/가장 긴 증가하는 부분 수열.java
import java.io.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(br.readLine());
        long[] numbers = new long[N + 1];
        long[] DP = new long[N + 1];
        String[] input = br.readLine().split(" ");
        for (int i = 1; i <= N; i++) {
            numbers[i] = Long.parseLong(input[i - 1]);
            DP[i] = 1;
        }

        long max = 1;
        for (int i = 2; i <= N; i++) {
            for (int j = 1; j < i; j++) {
                if (numbers[i] > numbers[j]) {
                    DP[i] = Math.max(DP[i], DP[j] + 1);
                }
            }
            
            max = Math.max(max, DP[i]);
        }

        System.out.println(max);
    }
}

File: 백준/Silver/11053. 가장 긴 증가하는 부분 수열/가장 긴 증가하는 부분 수열.js
let [n, input] = require("fs")
  .readFileSync(process.platform === "linux" ? "/dev/stdin" : "input.txt")
  .toString()
  .trim()
  .split("\n");

n = Number(n);
let numList = [0, ...input.split(" ").map(Number)];
let dp = Array(n + 1).fill(1);

let max = 1;
for (let i = 2; i <= n; i++) {
  for (let j = 1; j < i; j++) {
    if (numList[i] > numList[j]) {
      dp[i] = Math.max(dp[i], dp[j] + 1);
    }
  }

  max = Math.max(max, dp[i]);
}

console.log(max);
